### PR TITLE
Add processor for new spans schema

### PR DESCRIPTION
**Pull request by colin-sentry**
We are working on a new schema for spans, to better handle arbitrary keys (hashing), improve the sort key, etc.

This listens to the same kafka topic as the existing consumer, but puts data into the new schema.

### CODE REVIEW

File: rust_snuba/src/processors/mod.rs [hunk 1]
```diff
@@ -55,6 +56,7 @@ define_processing_functions! {
     ("QuerylogProcessor", "snuba-queries", ProcessingFunctionType::ProcessingFunction(querylog::process_message)),
     ("ReplaysProcessor", "ingest-replay-events", ProcessingFunctionType::ProcessingFunction(replays::process_message)),
     ("SpansMessageProcessor", "snuba-spans", ProcessingFunctionType::ProcessingFunction(spans::process_message)),
+    ("SpansV2MessageProcessor", "snuba-spans", ProcessingFunctionType::ProcessingFunction(spans_v2::process_message)),
```
Comment (colin-sentry) [inline rust_snuba/src/processors/mod.rs:59]: Is it fine if this has the same "key" as spans processor? It's unclear if it's just the kafka topic name - these two should have the same topic but different consumer groups

File: rust_snuba/src/processors/spans_v2.rs [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+            measurements.iter().for_each(|(k, v)| {
+                attr_num_buckets[(city_hash_64(k.as_bytes()) as usize) % attr_num_buckets.len()]
+                    .insert(k.clone(), v.value);
+                ()
+            });
+        }
+
+        Ok(res)
+    }
+}
```
Comment (colin-sentry) [inline rust_snuba/src/processors/spans_v2.rs:168]: will add tests

**Review by evanh — COMMENTED** [review]
This is missing tests but largely looks fine to me.

File: rust_snuba/src/processors/mod.rs (hunk 1 above)
Comment (untitaker) [reply rust_snuba/src/processors/mod.rs:59]: This is fine, check `snuba-generic-metrics`, it's basically the same setup.

File: rust_snuba/src/processors/spans_v2.rs [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+            span_id: u64::from_str_radix(&from.span_id, 16).unwrap_or_default(),
+            parent_span_id: from
+                .parent_span_id
+                .map_or(0, |s| u64::from_str_radix(&s, 16).unwrap_or(0)),
+            segment_id: from
+                .segment_id
+                .map_or(0, |s| u64::from_str_radix(&s, 16).unwrap_or(0)),
+            segment_name: sentry_tags.get("transaction").cloned().unwrap_or_default(),
+            is_segment: from.is_segment,
+            _sort_timestamp: (from.start_timestamp_ms / 1000) as u32,
```
[the remark this answers is not in the document]
Comment (colin-sentry) [reply rust_snuba/src/processors/spans_v2.rs:109]: might as well start with this and enhance with that later

File: rust_snuba/src/processors/spans.rs [hunk 4]
```diff
@@ -32,33 +32,33 @@ pub fn process_message(
 }
 
 #[derive(Debug, Default, Deserialize, JsonSchema)]
-struct FromSpanMessage {
-    description: Option<String>,
-    duration_ms: u32,
-    #[serde(alias = "end_timestamp_micro")]
```
Comment (Zylphrex) [inline rust_snuba/src/processors/spans.rs:38]: This alias looks like it disappeared

Comment (colin-sentry) [reply rust_snuba/src/processors/spans.rs:38]: that and the optional on timestamp_precise are both unused I think

File: rust_snuba/src/processors/spans_v2.rs [hunk 5]
```diff
@@ -0,0 +1,538 @@
...
+                &mut res.attr_str_44,
+                &mut res.attr_str_45,
+                &mut res.attr_str_46,
+                &mut res.attr_str_47,
+                &mut res.attr_str_48,
+                &mut res.attr_str_49,
+            ];
+
+            sentry_tags.iter().chain(tags.iter()).for_each(|(k, v)| {
+                attr_str_buckets[(half_md5(k.as_bytes()) as usize) % attr_str_buckets.len()]
```
Comment (phacops) [inline rust_snuba/src/processors/spans_v2.rs:373]: Are we sure we can use the same hashing mechanism on the querying side?

File: rust_snuba/src/processors/spans_v2.rs [hunk 6]
```diff
@@ -0,0 +1,538 @@
...
+    #[serde(skip_serializing_if = "HashMap::is_empty")]
+    attr_bool_9: HashMap<String, bool>,
+}
+
+fn half_md5(input: &[u8]) -> u64 {
+    let full_hash = md5::compute(input).0;
+    u64::from_be_bytes(full_hash[0..8].try_into().expect("should not happen"))
+}
+
+impl TryFrom<FromSpanMessage> for SpanV2 {
```
Comment (phacops) [inline rust_snuba/src/processors/spans_v2.rs:282]: If we're not going to throw any error, we can just use a `From` trait.

File: rust_snuba/src/processors/spans.rs [hunk 7]
```diff
@@ -32,33 +32,33 @@ pub fn process_message(
...
-    measurements: Option<BTreeMap<String, FromMeasurementValue>>,
-    parent_span_id: Option<String>,
-    profile_id: Option<Uuid>,
-    project_id: u64,
-    received: f64,
-    retention_days: Option<u16>,
-    segment_id: Option<String>,
-    sentry_tags: Option<BTreeMap<String, String>>,
-    span_id: String,
+pub(crate) struct FromSpanMessage {
```
Comment (phacops) [inline rust_snuba/src/processors/spans.rs:35]: Why do we need all this?

File: rust_snuba/src/processors/spans_v2.rs (hunk 5 above)
Comment (colin-sentry) [reply rust_snuba/src/processors/spans_v2.rs:373]: "half_md5" should be the hash with the best language support, it's in clickhouse too

File: rust_snuba/src/processors/spans.rs (hunk 7 above)
Comment (colin-sentry) [reply rust_snuba/src/processors/spans.rs:35]: spans and eap_spans both listen to the same kafka topic, so FromSpanMessage gets made crate public for use in both places

File: snuba/cli/devserver.py [hunk 8]
```diff
@@ -107,6 +107,17 @@ def devserver(*, bootstrap: bool, workers: bool) -> None:
                 *COMMON_RUST_CONSUMER_DEV_OPTIONS,
             ],
         ),
+        (
+            "spans-v2-consumer",
```
Comment (phacops) [inline snuba/cli/devserver.py:111]: `eap-spans-consumer`

File: rust_snuba/src/processors/eap_spans.rs [hunk 9]
```diff
@@ -0,0 +1,537 @@
...
+    // the span object for the new "events analytics platform"
+    organization_id: u64,
+    project_id: u64,
+    trace_id: Uuid,
+    span_id: u64,
+    #[serde(default)]
+    parent_span_id: u64,
+    segment_id: u64,      //aka transaction ID
+    segment_name: String, //aka transaction name
+    is_segment: bool,     //aka "is transaction"
```
Comment (phacops) [inline rust_snuba/src/processors/eap_spans.rs:41]: Those fields could be null in the future (span streaming) so let's add a `default` too.

File: rust_snuba/src/processors/eap_spans.rs [hunk 10]
```diff
@@ -0,0 +1,537 @@
...
+    _sort_timestamp: u32,
+    start_timestamp: u64,
+    end_timestamp: u64,
+    duration_ms: u32,
+    exclusive_time_ms: f64,
```
Comment (phacops) [inline rust_snuba/src/processors/eap_spans.rs:46]: Same thing.

File: rust_snuba/src/processors/eap_spans.rs [hunk 11]
```diff
@@ -0,0 +1,537 @@
...
+    retention_days: Option<u16>,
+    name: String, //aka description
```
Comment (phacops) [inline rust_snuba/src/processors/eap_spans.rs:48]: This is also an optional field.

Comment (colin-sentry) [reply rust_snuba/src/processors/eap_spans.rs:48]: this is handled in the trait:
`            name: from.description.unwrap_or_default(),`

**Review by untitaker — APPROVED** [review]
i assume storage is already merged

**Comment by untitaker** [comment]
for the test failures, run `cargo install cargo-insta` and `cargo insta test`

**Pull request merged**